File: cpp/algorithms/graph/scc/tarjan.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> tarjan_scc(const vector<vector<int>>& adj) {
    int n = (int)adj.size();
    vector<int> dfn(n, -1), low(n, 0), st;
    vector<char> on(n, 0);
    int timer = 0;
    vector<vector<int>> comps;
    function<void(int)> dfs = [&](int u) {
        dfn[u] = low[u] = timer++;
        st.push_back(u);
        on[u] = 1;
        for (int v : adj[u]) {
            if (dfn[v] == -1) {
                dfs(v);
                low[u] = min(low[u], low[v]);
            } else if (on[v])
                low[u] = min(low[u], dfn[v]);
        }
        if (low[u] == dfn[u]) {
            vector<int> comp;
            while (true) {
                int x = st.back();
                st.pop_back();
                on[x] = 0;
                comp.push_back(x);
                if (x == u) break;
            }
            comps.push_back(comp);
        }
    };
    for (int i = 0; i < n; ++i)
        if (dfn[i] == -1) dfs(i);
    return comps;
}

int main() {
    vector<vector<int>> g(4);
    g[0] = {1};
    g[1] = {2};
    g[2] = {0, 3};
    auto comps = tarjan_scc(g);
    assert(comps.size() == 2u);
    cout << "scc OK" << endl;
    return 0;
}

File: python/algorithms/graph/scc/tarjan.py
"""Tarjan 强连通分量。"""

from __future__ import annotations


def tarjan_scc(adj: list[list[int]]) -> list[list[int]]:
    n = len(adj)
    dfn = [-1] * n
    low = [0] * n
    st: list[int] = []
    on = [False] * n
    time = 0
    comps: list[list[int]] = []

    def dfs(u: int) -> None:
        nonlocal time
        dfn[u] = low[u] = time
        time += 1
        st.append(u)
        on[u] = True
        for v in adj[u]:
            if dfn[v] == -1:
                dfs(v)
                low[u] = min(low[u], low[v])
            elif on[v]:
                low[u] = min(low[u], dfn[v])
        if low[u] == dfn[u]:
            comp: list[int] = []
            while True:
                x = st.pop()
                on[x] = False
                comp.append(x)
                if x == u:
                    break
            comps.append(comp)

    for i in range(n):
        if dfn[i] == -1:
            dfs(i)
    return comps


if __name__ == "__main__":
    # 0->1->2->0 与 2->3
    g = [[] for _ in range(4)]
    g[0].append(1)
    g[1].append(2)
    g[2].extend([0, 3])
    comps = tarjan_scc(g)
    assert len(comps) == 2
    print("scc OK")
